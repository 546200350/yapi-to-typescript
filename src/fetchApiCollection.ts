import request from 'superagent'
import { ApiCollection, Config } from './types'

type LoginResult = {
  body: {
    errcode: number,
    errmsg: string,
  },
}
type ApiResult = { body: Buffer }

export default async function fetchApiCollection(config: Config): Promise<ApiCollection> {
  const [, baseUrl, projectId] = config.projectUrl.match(/(.+\/)project\/(\d+)\//)

  const loginUrl = `${baseUrl}api/user/${config.login.method === 'ldap' ? 'login_by_ldap' : 'login'}`
  const apiUrl = `${baseUrl}api/plugin/export?type=json&pid=${projectId}&status=all&isWiki=false`

  const agent = request.agent()

  if (config.extraCookies) {
    (agent as any).set('Cookie', config.extraCookies)
  }

  const { body: loginResult } = await agent
    .post(loginUrl)
    .send({
      email: config.login.email,
      password: config.login.password,
    }) as LoginResult
  if (!loginResult || loginResult.errcode !== 0) {
    throw new Error(
      `\x1b[31m登录 ${loginUrl} 失败，请检查邮箱、密码是否有误或服务是否可用。（服务器错误信息：${loginResult.errmsg}）\x1b[0m`
    )
  }

  const { body: api } = await agent.get(apiUrl).buffer() as ApiResult

  return JSON.parse(api.toString()) as any
}
