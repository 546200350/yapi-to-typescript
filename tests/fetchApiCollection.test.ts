import fetchApiCollection from '../src/fetchApiCollection'

test('登录失败', async () => {
  await expect(fetchApiCollection({
    projectUrl: 'http://foo.bar/project/20/interface/api',
    login: {
      email: 'hello@x.xx',
      password: 'errorPassword',
    },
  } as any)).rejects.toThrowErrorMatchingSnapshot()
})

test('登录成功，并成功获取 API 列表', async () => {
  expect(await fetchApiCollection({
    projectUrl: 'http://foo.bar/project/20/interface/api',
    login: {
      email: 'hello@x.xx',
      password: 'correctPassword',
    },
  } as any)).toMatchSnapshot()
})
