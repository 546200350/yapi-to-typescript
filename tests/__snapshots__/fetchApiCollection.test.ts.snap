// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`登录失败 1`] = `"登录 http://foo.bar/api/user/login 失败，请检查邮箱、密码是否有误或服务是否可用。"`;

exports[`登录成功，并成功获取 API 列表 1`] = `
Array [
  Object {
    "add_time": 1547006221,
    "desc": null,
    "index": 0,
    "list": Array [
      Object {
        "__v": 0,
        "_id": 45,
        "add_time": 1547006228,
        "api_opened": false,
        "catid": 58,
        "desc": "",
        "edit_uid": 0,
        "index": 0,
        "markdown": "",
        "method": "GET",
        "path": "/get",
        "project_id": 23,
        "query_path": Object {
          "params": Array [],
          "path": "/get",
        },
        "req_body_form": Array [],
        "req_body_is_json_schema": true,
        "req_headers": Array [],
        "req_params": Array [],
        "req_query": Array [],
        "res_body": "{\\"type\\":\\"object\\",\\"title\\":\\"empty object\\",\\"properties\\":{\\"data\\":{\\"type\\":\\"string\\"}},\\"required\\":[\\"data\\"]}",
        "res_body_is_json_schema": true,
        "res_body_type": "json",
        "status": "undone",
        "tag": Array [],
        "title": "get",
        "type": "static",
        "uid": 11,
        "up_time": 1547006250,
      },
      Object {
        "__v": 0,
        "_id": 48,
        "add_time": 1547006262,
        "api_opened": false,
        "catid": 58,
        "desc": "",
        "edit_uid": 0,
        "index": 0,
        "markdown": "",
        "method": "POST",
        "path": "/post",
        "project_id": 23,
        "query_path": Object {
          "params": Array [],
          "path": "/post",
        },
        "req_body_form": Array [],
        "req_body_is_json_schema": true,
        "req_body_other": "{\\"type\\":\\"object\\",\\"title\\":\\"empty object\\",\\"properties\\":{\\"page\\":{\\"type\\":\\"number\\"}},\\"required\\":[\\"page\\"]}",
        "req_body_type": "json",
        "req_headers": Array [
          Object {
            "_id": "5c35714d0ce3e84becbd13cc",
            "name": "Content-Type",
            "required": "1",
            "value": "application/json",
          },
        ],
        "req_params": Array [],
        "req_query": Array [],
        "res_body": "{\\"type\\":\\"object\\",\\"title\\":\\"empty object\\",\\"properties\\":{\\"list\\":{\\"type\\":\\"array\\",\\"items\\":{\\"type\\":\\"string\\"}}},\\"required\\":[\\"list\\"]}",
        "res_body_is_json_schema": true,
        "res_body_type": "json",
        "status": "undone",
        "tag": Array [],
        "title": "post",
        "type": "static",
        "uid": 11,
        "up_time": 1547006285,
      },
    ],
    "name": "test",
    "up_time": 1547006221,
  },
]
`;
